mpr-research-data.py

```diff
@@ -161,34 +122,156 @@ def sliceAndPushToGCP(courseDF, retrieveDF, client, targetBucketName=targetBucke
     return True
 
 
-# RETRIEVE KEYS
-# --------------------------------------------------------------------------
+class Config:
+    def __init__(self):
+        self.targetBucketName: str = 'mpr-research-data-uploads'
+        self.numberOfMonths: int = 4
+        self.defaultQueryFolder: str = 'queries'
+        self.queryTemplateDict: str = {'course': 'courseQuery.sql',
+                                       'retrieve': 'retrieveQuery.sql'}
+        self.dbParams: str = None
+        self.gcpParams: str = None
 
-dbParams = getDBCreds()
-gcpParams = getGCPCreds()
+    def set(self, name, value):
+        if name in self.__dict__:
+            self.name = value
+        else:
+            raise NameError('Name not accepted in set() method')
 
+    def setAndVerifySQLFile(self, queryType):
+        try:
+            self.queryTemplateDict[queryType] = str(
+                os.getenv(f'{queryType.upper()}_QUERY', self.queryTemplateDict[queryType]))
+
+            if self.queryTemplateDict[queryType].lower().endswith('.sql'):
+                logging.info(f'Loading in SQL file - {self.queryTemplateDict[queryType]}')
+            if not os.path.isfile(os.path.join(self.defaultQueryFolder, self.queryTemplateDict[queryType])):
+                logging.error(
+                    f'SQL Query file for {queryType} not found in query directory {self.defaultQueryFolder}.')
+                return False
+            return True
+        except Exception as e:
+            logging.error(f'Error Message: {e}')
+            logging.error(
+                f'Invalid parameter passed for query type: {queryType}.')
+            return False
+
+    def setAndVerifyDBCreds(self):
+
+        dbCredsDefaultDict = {'NAME': None,
+                              'USER': None,
+                              'PASSWORD': None,
+                              'HOST': None,
+                              'PORT': 3306}
+        self.dbParams = {}
+
+        allKeyPartsFound = True
+        for credPart in dbCredsDefaultDict:
+            try:
+                self.dbParams[credPart] = str(os.getenv(
+                    'DB_' + credPart, dbCredsDefaultDict[credPart]))
+            except Exception as e:
+                logging.error(f'Error Message: {e}')
+                logging.error(
+                    f'Invalid parameter passed for DB_{credPart}.')
+                allKeyPartsFound = False
+
+            if not self.dbParams[credPart]:
+                logging.error(
+                    f'Did not find .env variable for M-Write Peer Review production DB key: {credPart}.')
+                allKeyPartsFound = False
+
+        if not allKeyPartsFound:
+            return False
+
+    def setAndVerifyGCPCreds(self):
+        try:
+            self.gcpParams = json.loads(str(os.getenv('GCP_KEY')))
+            return True
+        except Exception as e:
```

Comment:
My assumption here for this approach is that we are trapping the error we think is most likely/expect to occur?

Reply:
As a follow-up, would an error in something such as failing to connect to the Database or GCP, where it could be more than just one expected error, do we try something else?

Reply:
Apologies for the late response…

Reply:
Apologies for the late response…

Yes, trap the exception that is most likely to occur.  In this case, as explained by the `json.loads()` documentation.  In testing, if we find that an additional exception occurs, we can add another `except` clause to handle it.  It's better to handle each exception individually than to catch every possible one with the generic `Exception`.

For connecting to the DB or GCP resources, catch the basic exceptions that their documentation says is likely to occur.  If we find other exceptions occur during testing, those could be handled with more `except` clauses.  If by "try something else" you mean trying to reconnect to the resource, I think that's not necessary.  It would be helpful to log messages about it, though.